// fivem-js/lib/enums/RadioStation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var RadioStation;
(function (RadioStation) {
    RadioStation[RadioStation["LosSantosRockRadio"] = 0] = "LosSantosRockRadio";
    RadioStation[RadioStation["NonStopPopFM"] = 1] = "NonStopPopFM";
    RadioStation[RadioStation["RadioLosSantos"] = 2] = "RadioLosSantos";
    RadioStation[RadioStation["ChannelX"] = 3] = "ChannelX";
    RadioStation[RadioStation["WestCoastTalkRadio"] = 4] = "WestCoastTalkRadio";
    RadioStation[RadioStation["RebelRadio"] = 5] = "RebelRadio";
    RadioStation[RadioStation["SoulwaxFM"] = 6] = "SoulwaxFM";
    RadioStation[RadioStation["EastLosFM"] = 7] = "EastLosFM";
    RadioStation[RadioStation["WestCoastClassics"] = 8] = "WestCoastClassics";
    RadioStation[RadioStation["BlaineCountyRadio"] = 9] = "BlaineCountyRadio";
    RadioStation[RadioStation["TheBlueArk"] = 10] = "TheBlueArk";
    RadioStation[RadioStation["WorldWideFM"] = 11] = "WorldWideFM";
    RadioStation[RadioStation["FlyloFM"] = 12] = "FlyloFM";
    RadioStation[RadioStation["TheLowdown"] = 13] = "TheLowdown";
    RadioStation[RadioStation["RadioMirrorPark"] = 14] = "RadioMirrorPark";
    RadioStation[RadioStation["Space"] = 15] = "Space";
    RadioStation[RadioStation["VinewoodBoulevardRadio"] = 16] = "VinewoodBoulevardRadio";
    RadioStation[RadioStation["SelfRadio"] = 17] = "SelfRadio";
    RadioStation[RadioStation["TheLab"] = 18] = "TheLab";
    RadioStation[RadioStation["RadioOff"] = 255] = "RadioOff";
})(RadioStation = exports.RadioStation || (exports.RadioStation = {}));
